### ISSUE TITLE
Use `multipleOutputPaths` as optimization

### ISSUE THREAD

**Issue by aameen-tulip** [stephank/yarn-plugin-nixify#20]
In my case populating the cache using `fetch-one` is painfully slow. ~6,000 dependencies, each takes ~0.3 ms to build with yarn, but ~2-4 seconds each spinning up build areas.

I think that by using multiple outputs in your cache builder you could drop a massive chunk of time spent spinning up and tearing down build areas.

Because the current cache builder processes these one at a time it takes ages.
By using flat hash you won't poison the hashes by changing to multiple outputs AFAIK.

Possibly an even faster way since the yarn plugin runs externally of `nix`: when populating the nix store from an existing cache, use `nix-store --add-fixed sha512 .yarn/cache/*` and modify the tarball name or something to match the expected store path name.

**Comment by adrian-gierakowski** [comment]
@aameen-tulip thank you for your suggestions. I’d be interested in improving this, however I’m not sure what you mean by  multipleOutputPaths?

**Comment by stephank** [comment]
> In my case populating the cache using `fetch-one` is painfully slow. ~6,000 dependencies, each takes ~0.3 ms to build with yarn, but ~2-4 seconds each spinning up build areas.
> 
> I think that by using multiple outputs in your cache builder you could drop a massive chunk of time spent spinning up and tearing down build areas.
> 
> Because the current cache builder processes these one at a time it takes ages. By using flat hash you won't poison the hashes by changing to multiple outputs AFAIK.

I agree this is a pain, and I've been thinking about adding a mode that simply builds the cache as a whole, instead of per dependency. (Like how `makeRustPackage` does it, if that sounds familiar. I believe other language solutions in Nixpkgs do similar stuff.)

I'm not sure how multiple outputs fit in, though. I'm not that familiar beyond how Nixpkgs uses them to split bin/lib/dev/doc. I guess the benefit is, even on rebuild, an intermediate Nix cache would only have to serve what has changed? Does that still require fixed output derivations, though, and how does that work with multiple outputs? (Or, I guess, the content-addressable stuff will solve that for us in the future.)

I always wonder if Nix handles 6000 outputs, or if there is some limit on it. 🙂

Aside from multiple outputs, I think the main implementation obstacle to doing a whole cache build is that we have to implement NAR and directory hashing on the JavaScript side, in order to create a fixed output derivation for the cache build. It's not super complicated, just a bunch of work.

> Possibly an even faster way since the yarn plugin runs externally of `nix`: when populating the nix store from an existing cache, use `nix-store --add-fixed sha512 .yarn/cache/*` and modify the tarball name or something to match the expected store path name.

I believe this is what preloading does? Or maybe I don't follow. 🙂
https://github.com/stephank/yarn-plugin-nixify/blob/9d735646df9f83e99936a1ecacc573356bc24f83/src/generate.ts#L296-L353
